src/OpenApi/Paths.hs

```diff
@@ -30,6 +41,79 @@ newtype OperationId = OperationId Text
   deriving newtype (JSON.FromJSON, ToText)
   deriving stock   (Eq, Show)
 
+data Responses = Responses
+  { default' :: Maybe Response
+  , patterns :: Map ResponseStatusPattern Response
+  }
+  deriving stock (Eq, Show)
+
+instance JSON.FromJSON Responses where
+  parseJSON = JSON.withObject "responses" $ \object -> do
+    default' <- object .:? "default"
+    patterns <- parsePatterns . HashMap.toList $ HashMap.delete "default" object
+    pure Responses{..}
+    where
+      parsePatterns
+        :: [(Text, JSON.Value)]
+        -> JSON.Parser (Map ResponseStatusPattern Response)
+      parsePatterns pairs = Map.fromList <$> traverse (uncurry parsePair) pairs
+
+      parsePair
+        :: Text
+        -> JSON.Value
+        -> JSON.Parser (ResponseStatusPattern, Response)
+      parsePair key value =
+        (,) <$> parseResponseStatusPattern key <*> JSON.parseJSON value
+
+      parseResponseStatusPattern :: Text -> JSON.Parser ResponseStatusPattern
+      parseResponseStatusPattern input =
+        either
+          (const . fail $ "Invalid status code pattern: " <> show input)
+          pure $ Text.parseOnly statusCodeParser input
+
+      statusCodeParser =
+        maybe (fail "Invalid status code") (pure . ResponseStatusExact)
```

Comment:
Not a blocker, but given `either` is on it's own line above, I would probably put `maybe` on it's own line too.

Reply:
Actually, I see now how the whole expression fits on one line, and the line below is the input feeding into the `maybe`. Please disregard.